File: leetcode/medium/74-searchA2DMatrix.js
// Write an efficient algorithm that searches for a value in an m x n matrix. This matrix has the following properties:

// Integers in each row are sorted from left to right.
// The first integer of each row is greater than the last integer of the previous row.
 
// Example 1:

// Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,50]], target = 3
// Output: true
// Example 2:
// Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,50]], target = 13
// Output: false
// Example 3:

// Input: matrix = [], target = 0
// Output: false

// first pass solution
var searchMatrix = function(matrix, target) {
    for (let i = 0; i < matrix.length; i++) {
        if (target >= matrix[i][0] && target <= matrix[i][matrix[i].length - 1]) {
            return binarySearch(matrix[i], target);
        }
    }
    return false;
}

var binarySearch = function(nums, target) {
    let low = 0;
    let high = nums.length - 1;
    
    while (low <= high) {
        let mid = Math.floor((low + high) / 2);
        
        if (nums[mid] === target) {
            return true;
        } else if (nums[mid] < target) {
            low = mid + 1;
        } else if (nums[mid] > target) {
            high = mid - 1;
        }
    }
    return false;
}

File: leetcode/medium/74-searchA2DMatrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int lowRow = 0;
        int highRow = matrix.size() - 1;

        /** binary search to find target row **/
        while (lowRow <= highRow) {
            int mid = floor((highRow + lowRow) / 2);
            int rowLen = matrix[mid].size() - 1;

            if (matrix[mid][0] <= target && matrix[mid][rowLen] >= target) {
                return binarySearch(matrix[mid], target);
            }

            if (target > matrix[mid][rowLen]) {
                lowRow = mid + 1;
            } else {
                highRow = mid - 1;
            }
        }

        return false;
    }
    /** binary search to find target inside of target row */ 
    bool binarySearch(vector<int>& vect, int target) {
        int low = 0;
        int high = vect.size() - 1;

        while (low <= high) {
            int mid = floor((high + low) / 2);

            if (vect[mid] == target) return true;

            if (vect[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }

        } 

        return false;
    }
};
